Use rust-analyzer from path if possible

WeetHet opened the pull request:
Release Notes:
- Added support for looking up the `rust-analyzer` binary in `$PATH`. This allows using such tools as `asdf` and nix to configure per-folder rust installations. To enable this behavior, use the `path_lookup` key when configuring the `rust-analyzer` `binary`: `{"lsp": {"rust-analyzer": {"binary": {"path_lookup": true }}}}`.

Question: Should we always pass the environment, no matter if the binary is pulled from settings or found in path?

WeetHet committed:
Use rust-analyzer from path if possible

SomeoneToIgnore commented on crates/languages/src/rust.rs:
```diff
@@ -35,29 +35,39 @@ impl LspAdapter for RustLspAdapter {
-                .collect(),
-            env: None,
-        })
+        } else {
```
We must `log::info!` or even `warn!` something here, can we do that in extensions?
I would log the error (a bit odd, that it was `ok()`-d before) or whatever other reason we're falling back to the path server, and the path itself (in case user's env is oddly configured and further debugging would be needed).

WeetHet replied:
Seems like it's not being done for Go:
https://github.com/zed-industries/zed/blob/16745542b585d9d70a04e298794ffc1e27ba687e/crates/languages/src/go.rs#L93-L102

SomeoneToIgnore reviewed (changes requested):
Would https://github.com/zed-industries/zed/pull/9293 help instead?
See also https://github.com/zed-industries/zed/issues/7902 for more general env-related plans.

Implicitly falling back to some randomly found server means not only convenience for somebody, but also makes things very hard to debug if things go wrong, IIRC, Zed does not anyhow indicate rust-analyzer's version used — so there's no way to understand in a particular moment in time which server is used with this PR merged.

So, not sure if it's a good way to go, I would ask @\mrnugget to review this if you insist — he's having the deepest context related to the Zed env plans.

jansol commented:
Particularly on NixOS this actually does not work for non-trivial projects. Non-trivial meaning that there are dependencies to system packages that may or may not be installed. Zed itself is one such project. To make sure rust-analyzer can find such dependencies (pkg-config, gcc, make, cmake, protoc, ...) it's necessary to use `nix develop --command rust-analyzer`, which in turn requires having a flake.nix at the root of each project.

My solution so far is to manually define it in my zed config:
```json

  "lsp": {
    "rust-analyzer": {
      "binary": {
        "path": "/usr/bin/nix",
        "arguments": ["develop", "--command", "rust-analyzer"]
      }
    },
  }
```

However for trivial self-contained projects this adds unnecessary friction with having to add a flake.nix and commit it to git. It would be nice to be able to specify an alternate behavior that just invokes rust-analyzer directly, without having to manually define it in a project-local zed config (which would defeat the point of not having to write a flake.nix).

SomeoneToIgnore commented:
> specify

I like the idea to be able to have it explicitly somehow, but opposed to the implicitness model in such cases.
I'm not very familiar with that or `asdf` tools, but I would imagine they are able to set extra env variables and launch arguments to Zed executable? 

It would be interesting to know what else is not working/missing/to-be-improved in Nix — ideally, as holistic as it was presented during Flatpak PRs, to see how many "hacks" are needed for Nix in general.

WeetHet commented:
> Would #9293 help instead? See also #7902 for more general env-related plans.
> 
> Implicitly falling back to some randomly found server means not only convenience for somebody, but also makes things very hard to debug if things go wrong, IIRC, Zed does not anyhow indicate rust-analyzer's version used — so there's no way to understand in a particular moment in time which server is used with this PR merged.
> 
> So, not sure if it's a good way to go, I would ask @\mrnugget to review this if you insist — he's having the deepest context related to the Zed env plans.

#9293 does not help, as right now, environment is not passed to rust-analyzer, so it fails, not being able to find cargo. This PR makes it so the environment would be passed. I've copied the code from the go-th one. I can try improving it by log-ing something on fail, but this kind of behaviour seem to be the default for other language's LSPs

WeetHet commented on crates/languages/src/rust.rs:
```diff
@@ -35,29 +35,39 @@ impl LspAdapter for RustLspAdapter {
+            let path = delegate.which(Self::SERVER_NAME.as_ref()).await?;
+            Some(LanguageServerBinary {
+                path,
+                arguments: vec![],
```
This does not seem to be the case for Go:
https://github.com/zed-industries/zed/blob/16745542b585d9d70a04e298794ffc1e27ba687e/crates/languages/src/go.rs#L98-L102

jansol commented:
Stuff being or not being present in the environment also affects other LSPs than rust-analyzer if they have dependencies on system packages. This will be especially important for future CMake / Meson integration (they have to detect dependency paths from the environment at runtime), but already affects clangd to an extent. Namely the problem be worked around with a `compile_commands.json` that contains full paths for dependency includes, but to generate that file you need to run cmake from a shell that does have the dependencies in the environment.

WeetHet commented:
> Stuff being or not being present in the environment also affects other LSPs than rust-analyzer if they have dependencies on system packages. This will be especially important for future CMake / Meson integration (they have to detect dependency paths from the environment at runtime), but already affects clangd to an extent. The problem be worked around with a `compile_commands.json` that contains full paths for dependency includes, but to generate that file you need to run cmake from a shell that does have the dependencies in the environment.

I think we just have to default to always passing the shell environment. That's a question for another issue though, as this pull request is just a quick and easy fix for me to be able to nuke rust from my global nix config

WeetHet commented:
> > I think we just have to default to always passing the shell environment. That's a question for another issue though, as this pull request is just a quick and easy fix for me to be able to nuke rust from my global nix config
> 
> _What_ shell environment? I launch Zed from the DE's application launcher.

Zed internally can launch your shell in the working directory to pick up stuff from direnv for example. It works great and I genuinely think passing that environment to the all language servers is a right thing to do, but again, out of scope of this PR

WeetHet committed:
Add an info notice when trying to use PATH

mrnugget commented:
I think the change here is correct and it's also what we do for other language servers (gopls, zls, ...). I even attempted this very same PR before, but I didn't merge it because it broke rust-analyzer in the `zed` repository for me and I didn't have enough time to figure out why that was. Does it work in this PR?

I have `rust-analyzer` in my PATH (I think from `rustup`?):

```
φ which rust-analyzer
/Users/thorstenball/.cargo/bin/rust-analyzer
```

And that didn't work with `zed` the last time I tried it.

So what I'm worried about is breaking rust-analyzer for Zed users because it behaves differently than gopls and others wrt to environment.

When I check out this PR and create a local bundle (`../script/bundle-mac -d -l`) so I can simulate the whole environment-being-picked-up-when-launched-from-Dock thing, `rust-analyzer` doesn't start for me in `zed`:

```
2024-06-03T11:11:09+02:00 [INFO] starting language server "rust-analyzer", path: "/Users/thorstenball/work/zed", id: 2
2024-06-03T11:11:09+02:00 [INFO] Trying to use `rust-analyzer` from PATH
2024-06-03T11:11:09+02:00 [INFO] found user-installed language server for Rust. path: "/Users/thorstenball/.cargo/bin/rust-analyzer", arguments: []
2024-06-03T11:11:09+02:00 [INFO] starting language server. binary path: "/Users/thorstenball/.cargo/bin/rust-analyzer", working directory: "/Users/thorstenball/work/zed", args: []
2024-06-03T11:11:09+02:00 [ERROR] crates/lsp/src/lsp.rs:360: cannot read LSP message headers
2024-06-03T11:11:09+02:00 [ERROR] crates/lsp/src/lsp.rs:782: server shut down
2024-06-03T11:11:09+02:00 [ERROR] crates/lsp/src/lsp.rs:382: Broken pipe (os error 32)
2024-06-03T11:11:09+02:00 [ERROR] failed to start language server "rust-analyzer": oneshot canceled
2024-06-03T11:11:09+02:00 [ERROR] server stderr: Some("error: 'rust-analyzer' is not installed for the toolchain '1.78-aarch64-apple-darwin'\n")
2024-06-03T11:11:09+02:00 [INFO] retrying installation of language server "rust-analyzer" in 1s
```

And there's no error shown anywhere, it just silently fails.

I think we need to fix that before we can merge it.

---

@jansol: Zed picks up your environment in different ways.

1. When Zed is launched from the CLI, it inherits the env.
2. When Zed is launched via your DE (i.e. clicking on Dock on macOSO), spawns a shell in your home directory to pick up the `env` there and use it.
3. When the language server adapter is configured to do so (which is what this is doing here), we spawn a shell in your project directory and pick up the shell env there, looking in _that_ PATH for the language server.

WeetHet commented:
> I think the change here is correct and it's also what we do for other language servers (gopls, zls, ...). I even attempted this very same PR before, but I didn't merge it because it broke rust-analyzer in the `zed` repository for me and I didn't have enough time to figure out why that was. Does it work in this PR?
> 
> I have `rust-analyzer` in my PATH (I think from `rustup`?):
> 
> ```
> φ which rust-analyzer
> /Users/thorstenball/.cargo/bin/rust-analyzer
> ```
> 
> And that didn't work with `zed` the last time I tried it.
> 
> So what I'm worried about is breaking rust-analyzer for Zed users because it behaves differently than gopls and others wrt to environment.
> 
> When I check out this PR and create a local bundle (`../script/bundle-mac -d -l`) so I can simulate the whole environment-being-picked-up-when-launched-from-Dock thing, `rust-analyzer` doesn't start for me in `zed`:
> 
> ```
> 2024-06-03T11:11:09+02:00 [INFO] starting language server "rust-analyzer", path: "/Users/thorstenball/work/zed", id: 2
> 2024-06-03T11:11:09+02:00 [INFO] Trying to use `rust-analyzer` from PATH
> 2024-06-03T11:11:09+02:00 [INFO] found user-installed language server for Rust. path: "/Users/thorstenball/.cargo/bin/rust-analyzer", arguments: []
> 2024-06-03T11:11:09+02:00 [INFO] starting language server. binary path: "/Users/thorstenball/.cargo/bin/rust-analyzer", working directory: "/Users/thorstenball/work/zed", args: []
> 2024-06-03T11:11:09+02:00 [ERROR] crates/lsp/src/lsp.rs:360: cannot read LSP message headers
> 2024-06-03T11:11:09+02:00 [ERROR] crates/lsp/src/lsp.rs:782: server shut down
> 2024-06-03T11:11:09+02:00 [ERROR] crates/lsp/src/lsp.rs:382: Broken pipe (os error 32)
> 2024-06-03T11:11:09+02:00 [ERROR] failed to start language server "rust-analyzer": oneshot canceled
> 2024-06-03T11:11:09+02:00 [ERROR] server stderr: Some("error: 'rust-analyzer' is not installed for the toolchain '1.78-aarch64-apple-darwin'\n")
> 2024-06-03T11:11:09+02:00 [INFO] retrying installation of language server "rust-analyzer" in 1s
> ```
> 
> And there's no error shown anywhere, it just silently fails.
> 
> I think we need to fix that before we can merge it.
> 
> @jansol: Zed picks up your environment in different ways.
> 
> 1. When Zed is launched from the CLI, it inherits the env.
> 2. When Zed is launched via your DE (i.e. clicking on Dock on macOSO), spawns a shell in your home directory to pick up the `env` there and use it.
> 3. When the language server adapter is configured to do so (which is what this is doing here), we spawn a shell in your project directory and pick up the shell env there, looking in _that_ PATH for the language server.

It seems that is due to zed having a rust-toolchain file, which specifies a toolchain for which rust-analyzer should be installed. I don't see any way to tell the user about it other than a notification and I have no idea how they work. So, what I'm thinking is that 
```
ERROR] server stderr: Some("error: 'rust-analyzer' is not installed for the toolchain '1.78-aarch64-apple-darwin'\n")
```
Should 100% be a notification and not just silently logged into the stderr

mrnugget commented:
> It seems that is due to zed having a rust-toolchain file, which specifies a toolchain for which rust-analyzer should be installed.

How/why does installing a global rust-analyzer work then?

> Should 100% be a notification and not just silently logged into the stderr

Yeah, the problem with that is that we retry a few times and this could get spammy quick.

WeetHet commented:
> How/why does installing a global rust-analyzer work then?

Hm, I guess one that rustup installes is a wrapper that tries to find a real one (similar to haskell-language-server for example), and the one we install is a real one

> Yeah, the problem with that is that we retry a few times and this could get spammy quick.

Could we do one notification like VSCode, where we say that language server has crashed multiple times, see logs for details

jansol commented:
Yeah, all tools installed by rustup are wrappers that select the right toolchain on a per-project basis. That was a bit of a surprise with cargo as well...

WeetHet commented:
> > Hm, I guess one that rustup installs is a wrapper that tries to find a real one (similar to haskell-language-server for example), and the one we install is a real one
> 
> Yeah. There's something going on that makes me scared of merging this. I think most people at Zed have `rust-analyzer` in `~/.cargo/bin` and if we merge this PR, we would essentially break Rust support for them, requiring manual deletion/reinstall of rust-analyzer, which is not ideal.
> 
> I'm wondering: should we make auto-detection a setting?

Yeah, sure why not. This would allow rustup based workflows to continue as expected while allowing local folder specific nix/asdf managed environments to work. I don't know how settings work though, so if you could help me out on that, it would be great (and would help me out later when I finally start working on better git selhosted support)

WeetHet committed:
Add a setting for detecting langserver from path

WeetHet commented:
@mrnugget could you check if this works for you (I don't have rustup)

WeetHet commented:
This PR should also have follow up PRs that modify other languages to respect `detect_path`

WeetHet commented:
@mrnugget how possible it is to merge this in before today's release? I'd love to have it in so I could use it. No rush though, I'm okay with waiting if you have anything else to deal with first

mrnugget commented:
@WeetHet sorry, I'm knees deep in X11 event loop stuff on Linux! I'll take a look tomorrow!

WeetHet commented:
> @WeetHet sorry, I'm knees deep in X11 event loop stuff on Linux! I'll take a look tomorrow!

Oh, okay, no worries, take your time, this is not that important

mrnugget commented on crates/project/src/project_settings.rs:
```diff
@@ -101,6 +101,9 @@ pub struct LspSettings {
     pub binary: Option<BinarySettings>,
```
I think we should make the new setting part of `binary` and name it `pathLookup` (it being `pathLookup: Option<bool>`). By default it's `None` and if it's `None`, then the language extension decides what the best default is (for `gopls`, for example, it does the lookup, for Rust it doesn't).

Users can then opt-in/out by setting it to `true`/`false`.

And if a `binary.path` is set, that overwrites the behaviour of `binary.pathLookup`.

WeetHet replied:
Okay, did just that

WeetHet committed:
Remove detect_path, add path_lookup to binary settings

If path_lookup is not set, we default to the language specific behaviour
Otherwise, if it's true, then we perform search for the binary in path,
if it's false, we don't. Currently this is only supported by
rust-analyzer

WeetHet committed:
Remove explicit ignore in C LSP adapter

mrnugget commented:
Changes work for me! Thanks. But I think we now need to modify the `go` and `c` implementations to check whether `path_lookup` isn't `Some(false)`, right?

WeetHet committed:
Respect path_lookup for c and go

WeetHet commented:
> Changes work for me! Thanks. But I think we now need to modify the `go` and `c` implementations to check whether `path_lookup` isn't `Some(false)`, right?

Sure thing, done!

mrnugget commented:
Okay, I tested this locally and everything seems to work as intended:

1. By default it uses the old way of using `rust-analyzer`
2. When I enable `path_lookup: true`, then it uses the local one defined through my `flake.nix`/direnv setup in `$PATH`

@SomeoneToIgnore since you have change requested here: okay if we merge this?

SomeoneToIgnore reviewed (approved):
Totally, almost all of my comments were implemented anyway 🙂
